chains/solana/contracts/programs/ccip-router/src/fee_quoter.rs

```diff
@@ -1,55 +1,225 @@
+use std::ops::AddAssign;
+
 use anchor_lang::prelude::*;
 use anchor_spl::{token::spl_token::native_mint, token_interface};
 use ethnum::U256;
 use solana_program::{program::invoke_signed, system_instruction};
 
 use crate::{
-    BillingTokenConfig, CcipRouterError, DestChain, Solana2AnyMessage, SolanaTokenAmount,
-    UnpackedDoubleU224, FEE_BILLING_SIGNER_SEEDS,
+    utils::{Exponential, Usd18Decimals},
+    BillingTokenConfig, CcipRouterError, DestChain, PerChainPerTokenConfig, Solana2AnyMessage,
+    SolanaTokenAmount, UnpackedDoubleU224, CCIP_LOCK_OR_BURN_V1_RET_BYTES,
+    FEE_BILLING_SIGNER_SEEDS,
 };
 
-// TODO change args and implement
+/// Any2EVMRampMessage struct has 10 fields, including 3 variable unnested arrays (data, receiver and tokenAmounts).
+/// Each variable array takes 1 more slot to store its length.
+/// When abi encoded, excluding array contents,
+/// Any2EVMMessage takes up a fixed number of 13 slots, 32 bytes each.
+/// For structs that contain arrays, 1 more slot is added to the front, reaching a total of 14.
+/// The fixed bytes does not cover struct data (this is represented by ANY_2_EVM_MESSAGE_FIXED_BYTES_PER_TOKEN)
+pub const ANY_2_EVM_MESSAGE_FIXED_BYTES: U256 = U256::new(32 * 14);
+
+/// Each token transfer adds 1 RampTokenAmount
+/// RampTokenAmount has 5 fields, 2 of which are bytes type, 1 Address, 1 uint256 and 1 uint32.
+/// Each bytes type takes 1 slot for length, 1 slot for data and 1 slot for the offset.
+/// address
+/// uint256 amount takes 1 slot.
+/// uint32 destGasAmount takes 1 slot.
+pub const ANY_2_EVM_MESSAGE_FIXED_BYTES_PER_TOKEN: U256 = U256::new(32 * ((2 * 3) + 3));
+
 pub fn fee_for_msg(
     _dest_chain_selector: u64,
     message: &Solana2AnyMessage,
     dest_chain: &DestChain,
-    token_config: &BillingTokenConfig,
+    fee_token_config: &BillingTokenConfig,
+    additional_token_configs: &[Option<BillingTokenConfig>],
+    additional_token_configs_for_dest_chain: &[PerChainPerTokenConfig],
 ) -> Result<SolanaTokenAmount> {
-    // TODO: Add all validations from lib.rs over the message here as well
-    message.validate(dest_chain, token_config)?;
-
-    let token = if message.fee_token == Pubkey::default() {
+    let fee_token = if message.fee_token == Pubkey::default() {
```

Comment:
We should have a check here for Chain family right? And only run this when the chain family is EVM